Source/Details/_ASDisplayViewAccessiblity.mm

```diff
@@ -109,64 +121,112 @@ - (CGRect)accessibilityFrame
   std::queue<ASDisplayNode *> queue;
-  queue.push(container);
+  queue.push(containerNode);
 
   // If the container does not have an accessibility label set, or if the label is meant for custom
   // actions only, then aggregate its subnodes' labels. Otherwise, treat the label as an overriden
   // value and do not perform the aggregation.
   BOOL shouldAggregateSubnodeLabels =
-      (container.accessibilityLabel.length == 0) ||
-      (container.accessibilityTraits & ASInteractiveAccessibilityTraitsMask());
+      (containerNode.accessibilityLabel.length == 0) ||
+      (containerNode.accessibilityTraits & ASInteractiveAccessibilityTraitsMask());
 
+  // Iterate through the whole subnode tree and aggregate
   ASDisplayNode *node = nil;
   while (!queue.empty()) {
     node = queue.front();
     queue.pop();
 
-    if (node != container && node.isAccessibilityContainer) {
-      UIView *containerView = node.isLayerBacked ? view : node.view;
-      CollectAccessibilityElementsForContainer(node, containerView, elements);
+    // If the node is an accessibility container go further down for collecting all the nodes
+    // information
+    if (node != containerNode && node.isAccessibilityContainer) {
+      UIView *view = containerNode.isLayerBacked ? containerView : containerNode.view;
+      AggregateSublabelsOrCustomActionsForContainerNode(node, view, elements);
       continue;
     }
 
+
+    // Aggregate either custom actions for specific accessibility traits or the accessibility labels
+    // of the node
     if (node.accessibilityLabel.length > 0) {
       if (node.accessibilityTraits & ASInteractiveAccessibilityTraitsMask()) {
         ASAccessibilityCustomAction *action = [[ASAccessibilityCustomAction alloc] initWithName:node.accessibilityLabel target:node selector:@selector(performAccessibilityCustomAction:)];
         action.node = node;
         [actions addObject:action];
 
+        // Connect the node with the custom action which representing it
         node.accessibilityCustomAction = action;
-      } else if (node == container || shouldAggregateSubnodeLabels) {
-        ASAccessibilityElement *nonInteractiveElement = [ASAccessibilityElement accessibilityElementWithContainer:view node:node];
+      } else if (node == containerNode || shouldAggregateSubnodeLabels) {
+        ASAccessibilityElement *nonInteractiveElement = [ASAccessibilityElement accessibilityElementWithContainerView:containerView node:node];
         [labeledNodes addObject:nonInteractiveElement];
+
+        // For ASTextNode accessibility container besides aggregating all of the
+        // accessibilityLabel's of the subnodes we are also collecting all of the link as
+        // custom actions
+        if (ASActivateExperimentalFeature(ASExperimentalTextNode2A11YContainer)) {
+          // Collect custom action for links
+          NSAttributedString *attributedText = nil;
```

Comment:
Can we hand off control here to the node class? Should we add API like `-accessibilityCustomActions` that ASTextNode can override?

Reply:
Agree. At the final stage we will give all nodes a way to provide "additional" `UIAccessibilityCustomActions` that is presenting additional custom actions. We cannot add `accessibilityCustomActions` as a new API though that get's bridged through to `UIView`. To not add more machinery around this for now I would suggest we roll with that what we have.